React/Views/RCTView.m

```diff
@@ -1563,4 +1564,28 @@ - (BOOL)performDragOperation:(id <NSDraggingInfo>)sender
 }
 #endif // ]TODO(macOS ISS#2323203)
 
+#pragma mark - Keyboard Events
+
+#if TARGET_OS_OSX // TODO: add iOS keyboard event handling
+- (void)keyDown:(NSEvent *)event {
+  if (!self.onKeyDown) {
+    [super keyDown:event];
+    return;
+  }
+
+  RCTViewKeyboardEvent *keyboardEvent = [RCTViewKeyboardEvent keyDownEventWithReactTag:self.reactTag characters:event.characters modifier:event.modifierFlags];
+  [_eventDispatcher sendEvent:keyboardEvent];
+}
+
```

Comment:
A common pattern with keyDown:/keyUp: events is for someone to override keyDown: check the type or keys pressed or other info on the event and optionally handle it, else pass the event up the responder chain by calling super. In this react implementation it looks like the presence of an onKeyDown is what determines whether the event gets handled, so how would someone implement a "conditional" key down handler, for example one that responds to arrow keys to move an object but lets other key events pass through to the responder chain to see if someone else handles it?

It looks like the current implementation would just drop all those events on the floor?

Reply:
Hmm good point. We could let the modules initialize this with the specific keys they want passed up and then not post events for all the other keys

Reply:
In [React] JS this would typically be controlled by conditionally invoking [`Event.prototype.stopPropagation()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation) or [`Event.prototype.preventDefault()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault), see e.g. https://reactjs.org/docs/events.html. I assume there’s precedent for this in the RN codebase, but haven’t had to implement this myself yet.

The only concern I’d have is that we can't get a synchronous response and, assuming that asynchronously responding to this event is ok, what the possible lag implications are when somebody's typing.

Reply:
@NickGerleman I assume this is a similar concern on Windows, no? For reference, the doc you linked to mentions:

> The properties in NativeSyntheticEvent like target, bubbles, cancelable etc., are also available for IKeyboardEvent and follow the same behaviors as other events in react-native today.

Reply:
I think @acoates-ms or @FalseLobster have context into this for Windows.

Reply:
Hello, I'm coming back to this thread almost 1 year later, but I had some concerns with the current implementation. As of right now, react-native-macos requires you pass 2 props to handle keyPress events: an `onKeyUp/Down` event handler, and a `validKeysUp/Down` array of strings. Every other platform just needs the former, and it's not at all obvious anywhere that the second prop is needed. Thinking of how to implement a conditional key handler, I came up with this:

**Proposal**: Let the native send the `onKeyDown` event up until it finds an RCTView that handles it. Then from there, let JS handle the propagation upwards. If I understand properly, these are bubbling events, so in JS they follow a similar propagation upwards until one of the superviews calls `event.stopPropogation` (like in this [FluentUI React Native code](https://github.com/microsoft/fluentui-react-native/blob/66dc9d3f6e33054604e7f1855a40ea1ea99ae745/packages/utils/interactive-hooks/src/useKeyCallback.win32.ts#L18)).

**Pros**: We don't need the extra prop, which is not well documented and a departure from cross-plat behavior. Also, more control is given in JS, aka the developer.

**Cons**: the onKeyUp/Down won't propagate past the React Root View. I think normally with the native responder chain, if an event isn't handled by any views, it goes up to the window, then app, then menus. Because my proposal is to pass it to JS (and not back to native if it reaches all the way up), nothing past the React Root View will get the event. 

I'm curious if there are glaring issues with the approach I just mentioned, or how windows handles the event propagation. I'm not at all familiar with how key events work there.

Reply:
It looks like react-native-windows does indeed have a similar prop: `keyDownEvents` https://github.com/microsoft/react-native-windows/pull/8077 .